sources/web/datalab/static/datalab.js

```diff
@@ -35,29 +35,29 @@ function initializeDataLab(
   var pageClass = document.body.className;
   if (pageClass.indexOf('notebook_app') >= 0) {
     notebookapp.preLoad(ipy, ipy.notebook, events, dialog, utils);
-    events.on('notebook_loaded.Notebook', function() {
+    // All extensions now load after the notebook which means we need to use promises
+    // instead of events. This may also affect the other events.on(...) calls below.
+    // See https://github.com/jupyter/notebook/issues/2499
+    promises.notebook_loaded.then(function() {
       notebookapp.postLoad(ipy, ipy.notebook, events, dialog, utils);
       window.datalab.loaded = true;
     });
-  }
-  else if (pageClass.indexOf('edit_app') >= 0) {
-    events.on('file_loaded.Editor', function() {
-      editapp.postLoad(ipy, ipy.editor);
+  } else if (pageClass.indexOf('edit_app') >= 0) {
+    $(document).ready(function() {
+      editapp.postLoad(ipy, ipy.editor, savewidget, events);
       window.datalab.loaded = true;
     });
-  }
-  else if (pageClass.indexOf('notebook_list') >= 0) {
+  } else if (pageClass.indexOf('notebook_list') >= 0) {
     // The page is finished loading after the notebook list is drawn for the first
     // time. The list is refreshed periodically though, so we need to only capture
     // the first occurrence
-    events.on('draw_notebook_list.NotebookList', function() {
+    promises.app_initialized.then(function() {
       if (!window.datalab.loaded) {
         notebooklist.postLoad(ipy.notebook_list, ipy.new_notebook_widget, dialog);
         window.datalab.loaded = true;
       }
     });
-  }
-  else if (pageClass.indexOf('session_list') >= 0) {
+  } else if (pageClass.indexOf('session_list') >= 0) {
     // The page is finished loading after the notebook list is drawn for the first
     // time. This event is used also after loading the terminal list. These lists are
     // refreshed periodically though, so we need to only capture the first occurrence
```

Comment:
No promise alternative for the line below? It'd be nice to get rid of `events` entirely as a dependency.

Reply:
I think `promises.app_initialized` should work, but even still we would need events for the savewidget in the editor